sync: introduce a thread-safe generic container and use it to remove a bunch of "GlobalMutex"es

vasild opened the pull request:
Introduce a generic container that provides a thread-safe access to any object by using a mutex which is acquired every time the object accessed.

For example:

```cpp
Synced<std::unordered_map<int, int>> m{{3, 9}, {4, 16}};

{
    SYNCED_LOCK(m, m_locked);

    // m_locked represents the internal object, i.e. std::unordered_map,
    // while m_locked is in scope the internal mutex is locked.

    auto it = m_locked->find(3);
    if (it != m_locked->end()) {
        std::cout << it->second << std::endl;
    }

    for (auto& [k, v] : *m_locked) {
        std::cout << k << ", " << v << std::endl;
    }
}

WITH_SYNCED_LOCK(m, p, p->emplace(5, 25));
```

Remove the global mutexes `g_maplocalhost_mutex`, `g_deadline_timers_mutex`, `cs_dir_locks`, `g_loading_wallet_mutex`, `g_wallet_release_mutex` and use `Synced<T>` instead.

## Benefits

_copied from a [comment below](https://github.com/bitcoin/bitcoin/pull/25390#issuecomment-1169788319):_

The `Synced<T>` abstraction is similar to what is suggested in [this comment](https://github.com/bitcoin/bitcoin/pull/24931#discussion_r890636987) but it does so in a generic way to avoid code repetition. Its benefit is:

1. It avoids code repetition at the implementation sites. See [PR#26151](https://github.com/bitcoin/bitcoin/pull/26151) for a live example. Namely this:

<details>
<summary>Lots of repetitions (92 lines)</summary>

```cpp
class Foo
{
public:
    void PushBack(x)
    {
        LOCK(m_mutex);
        m_data.push_back(x);
    }

    size_t Size()
    {
        LOCK(m_mutex);
        return m_data.size();
    }

    // maybe also other methods if needed...

    auto Lock()
    {
        return DebugLock<Mutex>{m_mutex, "Foo::m_mutex", __FILE__, __LINE__};
    }

private:
    Mutex m_mutex;
    std::vector<int> m_data;
};

class Bar
{
public:
    void PushBack(x)
    {
        LOCK(m_mutex);
        m_data.push_back(x);
    }

    size_t Size()
    {
        LOCK(m_mutex);
        return m_data.size();
    }

    // maybe also other methods if needed...

    auto Lock()
    {
        return DebugLock<Mutex>{m_mutex, "Bar::m_mutex", __FILE__, __LINE__};
    }

private:
    Mutex m_mutex;
    std::vector<std::string> m_data;
};

class Baz
{
public:
    void Insert(x)
    {
        LOCK(m_mutex);
        m_data.insert(x);
    }

    size_t Size()
    {
        LOCK(m_mutex);
        return m_data.size();
    }

    // maybe also other methods if needed...

    auto Lock()
    {
        return DebugLock<Mutex>{m_mutex, "Baz::m_mutex", __FILE__, __LINE__};
    }

private:
    Mutex m_mutex;
    std::set<std::string> m_data;
};
```
</details>

becomes this:

<details>
<summary>Short (3 lines)</summary>

```cpp
Synced<std::vector<int>> Foo;
Synced<std::vector<std::string>> Bar;
Synced<std::set<std::string>> Baz;
```
</details>

2. The mutex is properly encapsulated. With a global mutex and a global variable annotated with `GUARDED_BY()` it is indeed not possible to add new code that accesses the variable without protection (if using Clang and `-Wthread-safety-analysis` and `-Werror`), but it is possible to abuse the mutex and start using it to protect some more, possibly unrelated stuff (we already have this in the current code).

vasild committed:
sync: introduce a thread safe generic container

It can contain anything, encapsulates a mutex to synchronize access to
it and enfoces access through it.

vasild committed:
net: use Synced<T> for g_my_net_addr and remove g_maplocalhost_mutex

Convert `g_my_net_addr` to use `Synced<T>` and ditch the global mutex
`g_maplocalhost_mutex`.

w0xlt reviewed (commented):
Approach ACK

ryanofsky commented:
This seems like a potentially useful alternative to clang thread safety annotations. The `ThreadSafePtr<T>` class forces code to lock a mutex when accessing data, just like TSA annotations do, except unlike TSA annotations, it doesn't rely on a nonstandard compiler extension, or suffer from quirks that come from doing a limited static analysis.

`ThreadSafePtr<T>` is obviously not a complete substitute for thread safety annotations since it only handles the simple case where a single non-recursive Mutex is used to protect access to a single variable. But the variable can have any C++ type (primitive, container, or struct), so it's probably flexible enough for a lot of cases.

I don't think the `ThreadSafePtr<T>` is the best name because `xxx_ptr<T>` implies the type is some kind of lightweight reference to the `T`, not a container which holds the `T`. I would call it something like `Synced<T>` or `LockedData<T>`. (An analogy here would be `std::optional<T>`, which is not called `std::optional_ptr<T>`, even though it has `*` and `->` members, because it's purpose is to be a container, not a pointer. The class is named after what it's used for not what kind of members it has.)

ajtowns commented:
This seems like a non-starter to me? It's not even able to detect obvious double locks at compile-time:

```c++
    ThreadSafePtr<std::map<int, int>> m;
    m->emplace(5, 25);
    {
        auto m_locked = *m;
        m_locked->emplace(6, 36);
        m->emplace(7, 49);  // double lock

        auto m_locked2 = *m; // double lock
        m_locked->emplace(9, 81);
        m_locked2->emplace(10, 100);
    }
```

Even if it weren't worse at catching bugs, it doesn't seem like an improvement over writing:

```c++
    Mutex mut;
    std::map<int, int> m GUARDED_BY(mut);

    WITH_LOCK(mut, m.emplace(5, 25));
    {
        LOCK(mut);
        m.emplace(6, 36);
        m.emplace(7, 49);  // no double lock

        LOCK(mut); // double lock - detected at compile time
        m.emplace(9, 81);
        m.emplace(10, 100);
    }
```

ryanofsky commented:
> It's not even able to detect obvious double locks at compile-time: 

I assume it can be annotated just like any other lock. Agree implementation should fix this, though.

> Even if it weren't worse at catching bugs, it doesn't seem like an improvement over writing

Well one improvement is that it enforces locking on all compilers, unlike the clang annotations. The code itself doesn't seem much better or worse in this case for this very simple data structure, but it's probably is worth experimenting with for chains and chainstates and mempools, etc to be able to avoid logic bugs, distinguish different uses of cs_main for different purposes, and not just have LOCK(cs_main) everywhere what no indication about what exactly is being locked or why.

Looking at your examples, though I would even more want to replace the `*m` syntax with `m.Lock()` as suggested previously to make usage more obvious.

vasild commented on src/sync.h:
```diff
@@ -400,4 +400,98 @@ class CSemaphoreGrant
+    };
+
+    Proxy operator->() { return Proxy{m_mutex, m_obj}; }
+    const Proxy operator->() const { return Proxy{m_mutex, m_obj}; }
```
I removed them because they are not needed (currently). In a previous incarnation of this I had it working with `const ThreadSafePtr`, but removed the complications that were required for that.

Can always extend it (restore these) if `const` objects are necessary.

vasild commented on src/sync.h:
```diff
@@ -400,4 +400,98 @@ class CSemaphoreGrant
+        T* const m_raw_ptr_to_obj;
+    };
+
+    Proxy operator->() { return Proxy{m_mutex, m_obj}; }
+    const Proxy operator->() const { return Proxy{m_mutex, m_obj}; }
+
+    Proxy operator*() { return Proxy{m_mutex, m_obj}; }
+    const Proxy operator*() const { return Proxy{m_mutex, m_obj}; }
```
Renamed `operator*()` to `Lock()`. The `operator->()` is needed in order to cause a chained `->` calls.

Notice that the call:

```cpp
(*g_my_net_addr)[my_net_addr_entry] = lsi;
```

now becomes:

```cpp
g_my_net_addr.Lock()[my_net_addr_entry] = lsi;
```

vasild commented:
`e577177a33...a870b2b891`: address suggestions and remove optional commit that was a kind of scope creep for this PR - `net: simplify logic around reachable networks and -onlynet`, it will make it in a followup.

@ryanofsky very insightful review, thanks for the suggestions!

> This seems like a potentially useful alternative to clang thread safety annotations. The `ThreadSafePtr<T>` class forces code to lock a mutex when accessing data, just like TSA annotations do, except...

Hmm, right, I did not think of this from that perspective. In addition - TSA do not actually "force" anything, they emit a mere warning if compiled with `clang`. They do nothing for gcc. And if `-Werror` is not used to turn the warning into an error, then they can be missed/ignored.

I see this as a complementary to TSA.

> I don't think the `ThreadSafePtr<T>` is the best name because `xxx_ptr<T>` implies the type is some kind of lightweight reference to the `T`, not a container which holds the `T`. I would call it something like `Synced<T>` ...

Renamed to `Synced<T>`, thanks!

ajtowns commented:
> I assume it can be annotated just like any other lock. Agree implementation should fix this, though.

I think you'd have to mark the `Synced` object as being lock itself (`LOCKABLE`), and mark the `Proxy` class as being a RAII guard (`SCOPED_LOCKABLE`), with the constructor/destructor annotated appropriately (`EXCLUSIVE_LOCK_FUNCTION(ref_to_scoped_obj)` and `UNLOCK_FUNCTION()`), and with the functions that create the Proxy (`Lock` and `operator->`) annotated with negative constraints (`EXCLUSIVE_LOCKS_REQUIRED(!this)`)? I've got pretty low confidence that that will actually work as hoped/expected though...

> Well one improvement is that it enforces locking on all compilers, unlike the clang annotations.

We already enforce locking is correct via compiling with clang in CI; it's certainly an improvement to get those warnings earlier for anyone who's not using clang or doesn't have the options enabled, but

MarcoFalke commented:
I don't see any obvious benefit here, looking at the comparison with current code in https://github.com/bitcoin/bitcoin/pull/25390#issuecomment-1158434634

vasild committed:
scripted-diff: rename deadlineTimers to g_deadline_timers

-BEGIN VERIFY SCRIPT-
sed -i -e 's/deadlineTimers/g_deadline_timers/g' $(git grep -l deadlineTimers)
-END VERIFY SCRIPT-

vasild committed:
rpc: use Synced<T> for g_deadline_timers and remove g_deadline_timers_mutex

Convert `g_deadline_timers` to use `Synced<T>` and ditch the global mutex
`g_deadline_timers_mutex`.

vasild committed:
scripted-diff: rename dir_locks to g_dir_locks

-BEGIN VERIFY SCRIPT-
sed -i -e 's/\<dir_locks\>/g_dir_locks/g' $(git grep -l dir_locks)
-END VERIFY SCRIPT-

vasild committed:
rpc: use Synced<T> for g_dir_locks and remove cs_dir_locks

Convert `g_dir_locks` to use `Synced<T>` and ditch the global mutex
`cs_dir_locks`.

vasild committed:
wallet: use Synced<T> for g_loading_wallet_set and remove g_loading_wallet_mutex

Convert `g_loading_wallet_set` to use `Synced<T>` and ditch the global mutex
`g_loading_wallet_mutex`.

vasild committed:
wallet: use Synced<T> for g_unloading_wallet_set and remove g_wallet_release_mutex

Convert `g_unloading_wallet_set` to use `Synced<T>` and ditch the global mutex
`g_wallet_release_mutex`.

vasild committed:
scripted-diff: rename mapLocalHost to g_my_net_addr

-BEGIN VERIFY SCRIPT-
sed -i -e 's/mapLocalHost/g_my_net_addr/g' $(git grep -l mapLocalHost)
-END VERIFY SCRIPT-

vasild commented:
`a870b2b891...7b05e787cf`: use `LOCK(synced)` at call sites and remove more `GlobalMutex`es.

The benefit is not at the call sites - they can use different flavors of syntax sugar but all of them more or less boil down to the same thing. I changed it to use `{ LOCK(foo); foo->Method1(); foo->Method2(); ... }` so that it is not possible to misuse like @ajtowns suggested [above](https://github.com/bitcoin/bitcoin/pull/25390#issuecomment-1158434634).

The `Synced<T>` abstraction is similar to what is suggested in [this comment](https://github.com/bitcoin/bitcoin/pull/24931#discussion_r890636987) but it does so in a generic way to avoid code repetition. Its benefit is twofold:

1. It avoids code repetition at the implementation sites. Namely this:

<details>
<summary>Lots of repetitions</summary>

```cpp
class Foo
{
public:
    void PushBack(x)
    {
        LOCK(m_mutex);
        m_data.push_back(x);
    }

    size_t Size()
    {
        LOCK(m_mutex);
        return m_data.size();
    }

    // maybe also other methods if needed...

    auto Lock()
    {
        return DebugLock<Mutex>{m_mutex, "Foo::m_mutex", __FILE__, __LINE__};
    }

private:
    Mutex m_mutex;
    std::vector<int> m_data;
};

class Bar
{
public:
    void PushBack(x)
    {
        LOCK(m_mutex);
        m_data.push_back(x);
    }

    size_t Size()
    {
        LOCK(m_mutex);
        return m_data.size();
    }

    // maybe also other methods if needed...

    auto Lock()
    {
        return DebugLock<Mutex>{m_mutex, "Bar::m_mutex", __FILE__, __LINE__};
    }

private:
    Mutex m_mutex;
    std::vector<std::string> m_data;
};

class Baz
{
public:
    void Insert(x)
    {
        LOCK(m_mutex);
        m_data.insert(x);
    }

    size_t Size()
    {
        LOCK(m_mutex);
        return m_data.size();
    }

    // maybe also other methods if needed...

    auto Lock()
    {
        return DebugLock<Mutex>{m_mutex, "Baz::m_mutex", __FILE__, __LINE__};
    }

private:
    Mutex m_mutex;
    std::set<std::string> m_data;
};
```
</details>

becomes this:

<details>
<summary>Short</summary>

```cpp
Synced<std::vector<int>> Foo;
Synced<std::vector<std::string>> Bar;
Synced<std::set<std::string>> Baz;
```
</details>

2. The mutex is properly encapsulated. With a global mutex and a global variable annotated with `GUARDED_BY()` it is indeed not possible to add new code that accesses the variable without protection (if using Clang and `-Wthread-safety-analysis` and `-Werror`), but it is possible to abuse the mutex and start using it to protect some more, possibly unrelated stuff (we already have this in the current code).

MarcoFalke commented:
Ah, looks like this is `std::atomic` for structs/classes then

vasild commented:
> Ah, looks like this is `std::atomic` for structs/classes then

Right, kind of. `std::atomic` can be used for any trivially copyable structs/classes and if necessary it will use a mutex internally. The difference is that reading from an `atomic` would read it in a safe way from the memory and would return a copy of the stored object. So if we call a method of the stored object the mutex will be released before the method is called and the method will be called on the copy. `atomic` also does not provide a way to lock for longer time, spanning calls to multiple methods.

ryanofsky reviewed (commented):
I think right now I am at "concept -0.5" on this PR at 7b05e787cfc43494f0cea6364928f184c98e3fae. If other people think this change is a good idea, it seems fine, but I think the complexity it adds to sync.h isn't really justified by the minor code simplifications it enables elsewhere. Maybe if changes to `sync.h` were simplified or if a compelling use-case were explained, this would look like a better tradeoff.

---

re: https://github.com/bitcoin/bitcoin/pull/25390#issuecomment-1158434634

> This seems like a non-starter to me? It's not even able to detect obvious double locks at compile-time:

This is fixed now by allowing double locking. Following test passes `test_bitcoin -t sync_tests/synced_double_lock`

```diff
diff --git a/src/test/sync_tests.cpp b/src/test/sync_tests.cpp
index 55c2c5108de..3c225847529 100644
--- a/src/test/sync_tests.cpp
+++ b/src/test/sync_tests.cpp
@@ -140,4 +140,18 @@ BOOST_AUTO_TEST_CASE(inconsistent_lock_order_detected)
 #endif // DEBUG_LOCKORDER
 }
 
+BOOST_AUTO_TEST_CASE(synced_double_lock)
+{
+    Synced<std::map<int, int>> m;
+    m->emplace(5, 25);
+    {
+        auto m_locked = m.Lock();
+        m_locked->emplace(6, 36);
+        m->emplace(7, 49);  // double lock
+        auto m_locked2 = m.Lock(); // double lock
+        m_locked->emplace(9, 81);
+        m_locked2->emplace(10, 100);
+    }
+}
+
 BOOST_AUTO_TEST_SUITE_END()

```

vasild commented on src/sync.h:
```diff
@@ -400,4 +405,178 @@ class CSemaphoreGrant
+    {
+    public:
+        Proxy(Synced& parent, const char* mutex_name, const char* file_name, int line)
+            // Avoid double-lock if the current thread is already the owner.
+            // This read of parent.m_owner is unprotected but that is ok because
+            // the result of the comparison to the current thread cannot be
+            // changed by other, concurrently executing, threads.
+            : m_lock{parent.m_owner == std::this_thread::get_id() ? nullptr : &parent.m_mutex,
+                     mutex_name,
+                     file_name,
+                     line},
```
I agree. Simplified this by removing `m_owner` and using `RecursiveMutex`.

This reduced the `Synced` implementation from 110 to 58 lines (excluding comments).

It is possible to further reduce it by removing `Synced::UniqueLock` but then call sites would have to use `auto lock = foo.Lock();` instead of the sweet `LOCK(foo);`. I think either way is fine.

> Maybe it would be better to disallow double locking and use clang thread annotations to prevent obvious cases of double locking at compile time. But maybe this is not possible.

I will look at this again, but last time I tried I couldn't do that - stumbled on some limitations and bizarre behavior that looked like a bug in the thread safety annotations. Notice that even now some cases are prevented at compile time, like:

```cpp
LOCK(foo);
...
LOCK(foo);
```

vasild commented:
`7b05e787cf...acb21a5a24`: simplify the implementation of `Synced`, as suggested [above](https://github.com/bitcoin/bitcoin/pull/25390#discussion_r919129254).

vasild commented:
`acb21a5a24...4b870a1538`: rebase due to conflicts

vasild commented:
`4b870a1538...135ca49e24`: Further simplify the `Synced` class:

* Inherit `Synced` from `RecursiveMutex`. This allows the removal of the member `Synced::m_mutex` and the type `Synced::UniqueLock`
* Inherit `Synced::Proxy` from `::UniqueLock<RecursiveMutex>`. This allows the removal of the member `Synced::Proxy::m_lock`.

After the above changes there are two methods with similar names:
* `Synced::lock()`, inherited from `RecursiveMutex`, it just locks the object
* `Synced::Lock()`, it locks the object and returns a scoped lockable.

To avoid confusion, rename the latter to `operator*()`. This way constructs like `foo.Lock()[x]` become `(*foo)[x]` which looks better to me.

Now the `Synced` class is 47 lines (excluding comments).

vasild commented:
`135ca49e24...11ac578f91`: use `MutexType` instead of `M` in templates because `MutexType` is already used elsewhere in `sync.h`

ajtowns commented on src/sync.h:
```diff
@@ -400,4 +398,104 @@ class CSemaphoreGrant
+ * @endcode
+ */
+template <typename T>
+class Synced : public RecursiveMutex
```
I don't think we should be doing more recursive mutex stuff (they're harder to reason about, and less efficient any time that matters). I haven't been able to figure out a way of doing this sensibly otherwise, however -- clang just gets too confused about aliasing.

vasild replied:
It used `Mutex` before and mimicked a `RecursiveMutex`, but I totally agree with @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/25390#discussion_r919129254) and changed it to `RecursiveMutex`.

Even before that there was another approach which did not need a recursive mutex, and double locking was not allowed, but [you did not like that](https://github.com/bitcoin/bitcoin/pull/25390#issuecomment-1158434634).

ajtowns replied:
In my opinion: we shouldn't allow double locking (for almost all locks), and should give compile time errors when double locking is attempted. I'd love to see a variant of this that achieved both those things.

vasild replied:
Done - in the latest incarnation no double locking is allowed.

ajtowns reviewed (commented):
I'm still Approach NACK on this. :(

vasild commented on src/sync.h:
```diff
@@ -398,4 +398,104 @@ class CSemaphoreGrant
+     */
+    Proxy operator->()
+    {
+        return Proxy{*this, "Synced::operator->()", __FILE__, __LINE__};
```
Yeah, me too. Clang keeps emitting bogus warnings about loops like `for (auto& a : *obj)` for everything I tried, so I gave up. It has its limitations after all.

ajtowns replied:
Oh!! How about this:

```c++
template<typename T>
struct Guarded
{
private:
    mutable Mutex m;
    T obj GUARDED_BY(m);
public:
    class SCOPED_LOCKABLE Proxy
    {
    private:
        Guarded& g;
    public:
        Proxy(Guarded& g) EXCLUSIVE_LOCK_FUNCTION(g.m) : g{g}
        {
            g.m.lock();
        }
        ~Proxy() UNLOCK_FUNCTION() { g.m.unlock(); }

        T& operator*() { return g.obj; }
        T* operator->() { return &g.obj; }
    };
};

#define PROXY(name, guarded) decltype(guarded)::Proxy name{guarded};

Guarded<uint32_t> global_i;
Guarded<std::vector<uint32_t>> global_nums;
void foo()
{
    PROXY(p, global_i);
    *p = 5;
    *p += 6;
    // PROXY(p2, global_i); // error: acquiring mutex 'global_i.m' that is already held [-Werror,-Wthread-safety-analysis]
    // *p2 = 9;
}

int64_t lock_already_held(Guarded<std::vector<uint32_t>>::Proxy &p)
{
    int64_t j = 0;
    for (auto i : *p) { j += i; }
    return j;
}

int64_t bar()
{
    PROXY(p, global_nums);
    int64_t j = 0;
    for (auto i : *p) { j += i; }
    return j + lock_already_held(p);
}
```

That:

 * gives compile time errors if you try accessing without a lock
 * doesn't require recursive locks
 * gives clang compile time errors if you try double locking

which hits everything I wanted, and doesn't even seem too klunky. Perhaps needs some additional work to handle `const`.

ajtowns replied:
I think the macro is necessary -- otherwise `return Proxy{*this}` means you have to annotate the `Proxy(Proxy&&)` move constructor, but the only thing you can annotate that with is locking its internal reference to `g.m` but that then creates an alias of the mutex which prevents clang from detecting when double locking is happening since clang can't tell that two different invocations are locking aliases of the same lock.

Changing the name and the argument order makes lots of sense, I just posted it as soon as I found something that worked. `Guarded<T>` probably isn't a great name either. (I think it makes sense for `Proxy` to match whatever the macro is named)

vasild replied:
What about this:

```cpp
template <typename T>
class Synced
{
public:
    template <typename... Args>
    Synced(Args... args) : m_obj{args...}
    {
    }

    class SCOPED_LOCKABLE Proxy : ::UniqueLock<Mutex>
    {
    public:
        Proxy(Synced& parent) EXCLUSIVE_LOCK_FUNCTION(parent.m_mutex)
            : ::UniqueLock<Mutex>{parent.m_mutex, "...", "...", 123}, m_parent{parent}
        {
        }

        ~Proxy() UNLOCK_FUNCTION() {}

        T* operator->()
        {
            return &m_parent.m_obj;
        }

        template <typename Key>
        auto& operator[](Key key)
        {
            return (&m_parent.m_obj)->operator[](key);
        }

        auto begin()
        {
            return (&m_parent.m_obj)->begin();
        }

        auto end()
        {
            return (&m_parent.m_obj)->end();
        }

    private:
        Synced& m_parent;
    };

private:
    mutable Mutex m_mutex;
    T m_obj GUARDED_BY(m_mutex);
};

#define SYNCED_LOCK(synced, name) decltype(synced)::Proxy name{synced}
```

With example usage:

```cpp
Synced<std::vector<int>> synced{1, 2, 3, 4};

{
    SYNCED_LOCK(synced, p);

    p->push_back(5);
}

{
    SYNCED_LOCK(synced, p);

    std::cout << p[0] << std::endl;

    for (auto& i : p) {
        std::cout << i << std::endl;
    }
}
```

We loose the sweet syntax:

```cpp
Synced<std::vector<int>> synced{1, 2, 3, 4};

synced->push_back(5);
```

compared to the other variants, but this would still offer simplifications in higher level code and if you are ok with it, then I am fine.

ajtowns replied:
I'm not sure what the differences are between you're quoting above vs what I suggested? They currently look pretty much the same to me? (I was going to check that `begin`/`end`/`op[]` didn't create double locking issues, but you've gotten rid of them already)

I think it'd be worth adding `LIFETIMEBOUND` annotations (from attributes.h) as well. Perhaps you should delete the move constructor for `Synced`?

I agree `synced->push_back(5)` is pretty, but it seems likely to be bug prone to me: consider `it = synced_map->find(key)` -- the lock will be released as soon as the iterator is returned, allowing `it` to be immediately invalidated. So I kind of think we should just count ourselves lucky that we can't make it work and callers have to be explicit about the lock scope...

ajtowns replied:
It might be nice to have a `ConstProxy` variant too, so that if you have a `const` reference to an object containing a `Synced<foo>` you can still access have synced access to the object as `const foo`. On the other hand, maybe better to worry about that when an actual use comes up.

ajtowns replied:
`Proxy : ::UniqueLock<Mutex>` -- isn't it better to say `private` explicitly? (Guess who just assumed it defaulted to `public` until the compiler told him otherwise...)

vasild replied:
> I'm not sure what the differences are between you're quoting above vs what I suggested?

Yes, they are pretty much the same. The question was maybe more directed to @ryanofsky - if he is ok with that. I want to avoid pushing another iteration of this PR and it to be shot by somebody shortly after :) If you both agree on the above, then lets do that!

To-polish: LIFETIMEBOUND, delete the Synced move constructor, consider const, explicit inheritance type.

ajtowns replied:
If you made `m_obj` and `m_mutex` be protected instead of private, that would allow you to write:

```c++
class MyFoo : public Synced<foo>
{
public:
    int pretty() { LOCK(m_mutex); return m_obj.pretty(); }
};
```

and then do `i = myfoo.pretty();` -- gets you the same end result as the `synced->pretty()` stuff, but only for things you explicitly add a wrapper for, which at least also gives you a chance to make sure it's actually safe to be used in isolation?

vasild replied:
Ok, I wouldn't add features that are not used anywhere, but `s/private/protected` is not actually more code. Makes sense if somebody wants to inherit it and do more involved things.

To-polish: LIFETIMEBOUND, delete the Synced move constructor, consider const, explicit inheritance type, protected `m_mutex` and `m_obj`.

@ryanofsky, what do you think?

ryanofsky replied:
> To-polish: LIFETIMEBOUND, delete the Synced move constructor, consider const, explicit inheritance type, protected `m_mutex` and `m_obj`.
> 
> @ryanofsky, what do you think?

All of these tweaks sound good to me.

Generally speaking about the PR as a whole, I'm a skeptical the `Synced` class will be broadly useful, but I think it's a simple enough utility that can be used in a few places without many drawbacks, and that it's worth experimenting with. I was more interested in the class previously when I thought it was a locking pointer enforcing use of an external mutex (that could be used to migrate code away from using `cs_main`), than a locking container that just used an internal mutex. But it seems like a good thing to use in places where it fits.

vasild replied:
Updated, I only left the `const` stuff away, since it is not needed now and can be added later if needed.

For the usefulness of the `Synced` class, see:
* the second half of https://github.com/bitcoin/bitcoin/pull/25390#issuecomment-1169788319
* the second half of https://github.com/bitcoin/bitcoin/pull/26151#pullrequestreview-1123626545
* the commits from this PR that put it into action.

ajtowns replied:
Hmm, when used with shared locks, it seems like clang's thread safety analysis doesn't actually prevent you from calling non-const member functions of a `GUARDED_BY` object when only holding the lock in shared mode (it does prevent you from doing `a = b`, just not `a.set(b)`). So if you take a shared lock over a vector, then resize it and invalidate everyone elses iterators, clang thinks it's all fine.

Which I think means that if we wanted to do shared locking more broadly, then this method might be the safest way we have of it: ie have `SYNCED_SHARED(foo,  fooproxy)` take a shared lock and have `*fooproxy` give a `const Foo&` so you can't accidently modify `foo` while you're potentially sharing it with other threads.

vasild commented:
`11ac578f91...70dac32e76`: rebase due to conflicts + update as per the discussion in https://github.com/bitcoin/bitcoin/pull/25390#discussion_r978269158

ajtowns reviewed (commented):
Approach ACK

Don't think the bool-array to unordered_set makes much sense, otherwise looks good.

vasild commented:
`70dac32e76...c3e4c34219`: rebase and address suggestions

vasild commented on src/net.cpp:
```diff
@@ -325,14 +330,17 @@ void SetReachable(enum Network net, bool reachable)
 {
-    LOCK(g_maplocalhost_mutex);
-    return !vfLimited[net];
+    return WITH_SYNCED_LOCK(g_reachable_networks, p, return p->count(net) > 0);
```
Done.

vasild commented on src/sync.h:
```diff
@@ -397,4 +398,77 @@ class CSemaphoreGrant
+    T m_obj GUARDED_BY(m_mutex);
+};
+
+#define SYNCED_LOCK(synced, name) decltype(synced)::Proxy name{synced, #synced, __FILE__, __LINE__}
```
Sometimes the hardcoded name could become too long/verbose for a too short scope: e.g. `g_reachable_networks_proxy` or `g_reachable_networks_locked` when it is just used on the line below and nowhere else.

I added now the possibility to call `SYNCED_LOCK()` with 1 or 2 arguments and opted for a `_locked` suffix instead of `_proxy` for the 1-arg case. So the callers have complete flexibility on how to call it.

vasild replied:
That does not compile on windows: https://cirrus-ci.com/task/4559637407072256?logs=build#L814 :-/

If you are ok with this, I will fall back to the 2-arg macro. Then the caller can pick the name depending on their preferences - `p`, `foo_proxy`, `foo_locked` or whatever.

ajtowns replied:
If you want naming it to be optional, probably better to have different macros, same as `LOCK(m)` and `WAIT_LOCK(m,g)`. I'm not really sure what I prefer here.

vasild replied:
I went back to what it was before - `SYNCED_LOCK(object, name)`. The 1-arg macro was used in just one place and it did not compile on windows, not worth it. By selecting the name the caller can choose whether to use a short one for small scopes of a few lines, or a longer and more descriptive one for bigger scopes.

vasild commented on src/net.cpp:
```diff
@@ -115,7 +117,10 @@ bool fDiscover = true;
-static bool vfLimited[NET_MAX] GUARDED_BY(g_maplocalhost_mutex) = {};
+
+static Synced<std::unordered_set<Network>> g_reachable_networks{
+    NET_UNROUTABLE, NET_IPV4, NET_IPV6, NET_ONION, NET_I2P, NET_CJDNS, NET_INTERNAL};
```
My reasoning is that the most natural way to express this is to have a list of networks that are reachable (or unreachable if the logic is to be inverted) and `std::set` is the best way to do that. Having an array of `bool`s is a bit hackish as it requires that all enums are sequential and start at `0` and requires the existence of `NET_MAX` which smells bad. My plan is to remove `NET_MAX`. Ideally, enum lables should be used without depending on their specific integer values.

Anyway, if this is controveral I will ditch that change and leave it as `vfLimited[NET_MAX]` since it is not the main topic of this PR. Should I?

ajtowns replied:
enums are sequential and automatically start at 0; and `NET_MAX` is added for precisely that purpose. That's the way these things are designed to be used, it's not hackish or a bad smell...

vasild replied:
I disagree, but that discussion is for another PR. I reverted it back to what it is in `master` - an array of `bool`s.

vasild commented on src/net.cpp:
```diff
@@ -115,7 +117,10 @@ bool fDiscover = true;
+
+static Synced<std::unordered_set<Network>> g_reachable_networks{
+    NET_UNROUTABLE, NET_IPV4, NET_IPV6, NET_ONION, NET_I2P, NET_CJDNS, NET_INTERNAL};
+
```
> ... separate mutexes ...

What you propose will use just one mutex? Anyway, the list of reachable networks is unrelated to the list of local addresses. I do not think they should be grouped into one `struct` or protected by the same mutex.

ajtowns replied:
Yes, it was an "alternative .. to .. separate mutexes". Mostly intended to show using this pattern doesn't require moving everything to a dedicated mutex if that's not desired. In this case, like I said "hard to see it making much practical difference" though.

vasild replied:
Ok, I am leaving it as it is and marking this as resolved. I think it is natural to guard unrelated things with separate mutexes.

vasild commented:
`0f457d2bdd...6c0bcd0928`: update comment

vasild commented:
`6c0bcd0928...f8f7e62f81`: don't change the type of the variable that holds the list of reachable networks (an array of `bool`s), see https://github.com/bitcoin/bitcoin/pull/25390#discussion_r998202653

vasild commented:
`f8f7e62f81...3db109e794`: rebase due to conflicts

@ajtowns, I think I addressed all your suggestions?

vasild commented:
`3db109e794...e42ce20c65`: rebase due to conflicts

vasild commented:
`e42ce20c65...ebfe47e594`: rebase due to conflicts

jonatack reviewed (commented):
Light initial ACK ebfe47e5942fa5e0ce21b44511258b6dfa4cf621 modulo poking around with the thread safety coverage/guarantees to break them. Compiled and ran unit tests at each commit with ARM64 clang 16 (except the scripted-diff ones).

Feel free to ignore the comments below unless you need to repush.

vasild commented on src/sync.h:
```diff
@@ -397,4 +398,77 @@ class CSemaphoreGrant
+    T m_obj GUARDED_BY(m_mutex);
+};
+
+#define SYNCED_LOCK(synced, name) decltype(synced)::Proxy name{synced, #synced, __FILE__, __LINE__}
```
Named args at call sites look too verbose to me and here a short `name` is ok because the scope if very limited - only used on one line.

vasild commented:
`ebfe47e594...38c0d628e0`: rebase due to conflicts

@hebasto, if the purpose of the low level code is to make the higher level code more robust and less bug-prone, then I think `Synced` achieves this. The pattern:

```cpp
Mutex m;
Type my_variable GUARDED_BY(m);
```

is very common. The `Synced` structure encapsulates both, enforces the idiom and makes it impossible to "to abuse the mutex and start using it to protect some more, possibly unrelated stuff".

It also helps remove `GlobalMutex` usages. `GlobalMutex` could be [confusing](https://github.com/bitcoin/bitcoin/pull/27912#discussion_r1259704670).

vasild commented:
`38c0d628e0...b749c4d2ca`: rebase due to conflicts

vasild commented:
`b749c4d2ca...92a444d0d7`: rebase due to conflicts, drop commit 93e60b9a0c3ee44fbf1f64c91e908c3e507d1d7f `net: detach vfLimited from g_maplocalhost_mutex` because `vfLimited` + `SetReachable()` have already been encapsulated (see 6e308651c441cbf8763c67cc099c538c333c2872 from #27071).

vasild commented:
`92a444d0d7...f3489638dc`: rebase due to conflicts

maflcko reviewed (commented):
Are there any examples from the last couple of years where this would have helped to prevent a  bug or other issue?

vasild commented:
> Are there any examples from the last couple of years where this would have helped to prevent a bug or other issue?

Here is an example where two unrelated variables were guarded by the same mutex (already fixed in the latest code):

https://github.com/bitcoin/bitcoin/blob/c42ded3d9bda8b273780a4a81490bbf1b9e9c261/src/net.cpp#L116-L118

vasild commented:
Closing this due to no interest from reviewers for a long time plus currently my hands are full with more important things. Would be happy to revisit if there is interest.
